src/libopensc/card-jacartapki.c

```diff
@@ -0,0 +1,1905 @@
+
+	return SC_SUCCESS;
+}
+
+static int
+jacartapki_init(struct sc_card *card)
+{
+	struct sc_context *ctx = card->ctx;
+	struct jacartapki_private_data *private_data = NULL;
+	struct sc_path path;
+	unsigned int flags;
+	int rv = SC_ERROR_NO_CARD_SUPPORT;
+
+	LOG_FUNC_CALLED(ctx);
+
+	private_data = (struct jacartapki_private_data *)calloc(1, sizeof(struct jacartapki_private_data));
+	if (private_data == NULL)
+		LOG_ERROR_RET(ctx, SC_ERROR_OUT_OF_MEMORY, "Failed to allocate private blob for card driver.");
+
+	memset(private_data, 0, sizeof(struct jacartapki_private_data));
+
+	private_data->auth_state[0].pin_reference = JACARTAPKI_USER_PIN_REFERENCE;
+	private_data->auth_state[1].pin_reference = JACARTAPKI_SO_PIN_REFERENCE;
+
+	card->cla = 0x00;
+	card->drv_data = private_data;
+
+	sc_path_set(&path, SC_PATH_TYPE_DF_NAME, jacartapki_aid.value, jacartapki_aid.len, 0, 0);
+	rv = sc_select_file(card, &path, NULL);
+	if (rv < 0) {
+		free(card->drv_data);
+		card->drv_data = NULL;
+		LOG_ERROR_RET(ctx, SC_ERROR_INVALID_CARD, "Failed to select JaCarta PKI AID.");
+	}
+
+	rv = jacartapki_get_serialnr(card, NULL);
+	LOG_TEST_GOTO_ERR(ctx, rv, "Cannot get card serial");
+
+	rv = jacartapki_get_caps(card);
+	LOG_TEST_GOTO_ERR(ctx, rv, "Cannot get card capabilities");
+
+	flags = JACARTAPKI_CARD_DEFAULT_FLAGS;
+	_sc_card_add_rsa_alg(card, 1024, flags, 0x10001);
+	_sc_card_add_rsa_alg(card, 2048, flags, 0x10001);
+	_sc_card_add_rsa_alg(card, 4096, flags, 0x10001);
+
+	card->caps |= SC_CARD_CAP_RNG;
+	card->caps |= SC_CARD_CAP_APDU_EXT;
+
+	rv = jacartapki_load_options(card);
+	LOG_TEST_GOTO_ERR(ctx, rv, "Failed to read card driver configuration");
+
+#if defined(ENABLE_SM)
+	card->sm_ctx.ops.open = jacartapki_iso_sm_open;
+	card->sm_ctx.ops.get_sm_apdu = jacartapki_iso_sm_get_apdu;
+	card->sm_ctx.ops.free_sm_apdu = jacartapki_iso_sm_free_apdu;
+
+#endif
+	rv = SC_SUCCESS;
+err:
+	LOG_FUNC_RETURN(ctx, rv);
+}
+
+static int
+jacartapki_read_binary(struct sc_card *card, unsigned int offs,
+		u8 *buf, size_t count, unsigned long *flags)
+{
+	struct sc_context *ctx = card->ctx;
+	struct sc_apdu apdu;
+	int rv;
+	const size_t leMax = 0x100U;
+	size_t binaryDataRead;
+
+	LOG_FUNC_CALLED(ctx);
+	sc_log(ctx, "jacartapki_read_binary(card:%p) offs %i; count %" SC_FORMAT_LEN_SIZE_T "u", card, offs, count);
+	if (offs > 0x7fff) {
+		sc_log(ctx, "invalid EF offset: 0x%X > 0x7FFF", offs);
+		return SC_ERROR_OFFSET_TOO_LARGE;
+	}
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, 0xB0, (offs >> 8) & 0x7F, offs & 0xFF);
+	apdu.le = MIN(count, leMax);
+	apdu.resplen = count;
+	apdu.resp = buf;
+#ifdef ENABLE_SM
+	if (card->sm_ctx.sm_mode != SM_MODE_NONE)
+		apdu.flags |= SC_APDU_FLAGS_NO_GET_RESP;
+#endif
+
+	rv = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(ctx, rv, "APDU transmit failed");
+	binaryDataRead = apdu.resplen;
+	if (apdu.sw1 == 0x61) {
+		/* usually we get here when SM is on */
+		unsigned char getRespBuf[0x100];
+		while (binaryDataRead < count) {
```

Comment:
Do you want to throw an error if the bytes available via GET RESPONSE are bigger than the output buffer?